### ISSUE TITLE
train with multi gpus, always blocking...

### ISSUE THREAD

**Issue by GeorgePearse** [BinItAI/visdet#877]
**Original issue:** https://github.com/open-mmlab/mmdetection/issues/8265
**Created:** 2022-06-25
**Last updated:** 2022-06-26

---

Train with 1 node, 8 gpus is ok. But train with 2 nodes, 16 gpus is not ok.

1. train with 2 nodes(a & b), each node has 8 gpus.
2. run with the following shell scripts：./tools/dist_train.sh configs/goods/faster_rcnn_r50_fpn_2x_coco.py 8

in dist_train.sh:
node a is :
#!/usr/bin/env bash

CONFIG=$1
GPUS=$2
NNODES=${NNODES:-2}
NODE_RANK=${NODE_RANK:-0}
PORT=${PORT:-29511}
MASTER_ADDR=${MASTER_ADDR:-"10.193.21.228"}

PYTHONPATH="$(dirname $0)/..":$PYTHONPATH \
python -m torch.distributed.launch \
    --nnodes=$NNODES \
    --node_rank=$NODE_RANK \
    --master_addr=$MASTER_ADDR \
    --nproc_per_node=$GPUS \
    --master_port=$PORT \
    $(dirname "$0")/train.py \
    $CONFIG \
    --seed 0 \
    --launcher pytorch ${@:3}

node b is:

#!/usr/bin/env bash

CONFIG=$1
GPUS=$2
NNODES=${NNODES:-2}
NODE_RANK=${NODE_RANK:-1}
PORT=${PORT:-29511}
MASTER_ADDR=${MASTER_ADDR:-"10.193.21.228"}

PYTHONPATH="$(dirname $0)/..":$PYTHONPATH \
python -m torch.distributed.launch \
    --nnodes=$NNODES \
    --node_rank=$NODE_RANK \
    --master_addr=$MASTER_ADDR \
    --nproc_per_node=$GPUS \
    --master_port=$PORT \
    $(dirname "$0")/train.py \
    $CONFIG \
    --seed 0 \
    --launcher pytorch ${@:3}

But, the programming will block and print ：
/mnt/anaconda3/envs/openmmlab/lib/python3.8/site-packages/torch/distributed/launch.py:178: FutureWarning: The module torch.distributed.launch is deprecated
and will be removed in future. Use torchrun.
Note that --use_env is set by default in torchrun.
If your script expects `--local_rank` argument to be set, please
change it to read from `os.environ['LOCAL_RANK']` instead. See
https://pytorch.org/docs/stable/distributed.html#launch-utility for
further instructions

  warnings.warn(
WARNING:torch.distributed.run:
*****************************************
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed.
*****************************************

Please help me.
